### ISSUE TITLE
Leave one dataset out cross validation

### ISSUE THREAD

**Issue by MehriB** [MIC-DKFZ/nnUNet#348]
Hi Fabian,

I have data from 7 different datasets. And I would like to use nnunet for my segmentation but I want to modify the cross validation. 
Instead of random 5-fold validation. I would like to have leave one dataset out cross validation? so train on 6 different networks and each time one of the datasets being used for the cross validation.
Does nnunet offer an option for this? or I have to change some files manually? like the cropping.py?

Thank you for the help,
Mehri

**Comment by FabianIsensee** [comment]
Hi,
you have two options:

1) (quick and dirty) merge all the training cases into one big dataset. Then, after `nnUNet_plan_and_preprocess`, create your own split file (see documentation) so that the datasets are split appropriately. This is dirty because the experiment planning will have access to the metadata of all cases
2) (clean but more work) create 7 new datasets (each of them is a merger of 6 of your datasets. One dataset is left out). You can run `nnUNet_plan_and_preprocess` independently for each of them and then after the training is done use `nnUNet_predict` to get predictions on the held out dataset.

Best,
Fabian

**Comment by MehriB** [comment]
Thanks for the help.
I think I will try option 1, as I want the network to improve by the left out dataset. 
Option 2 seems that the left out data set will be used as a test set not as the cross validation set. 

One question, if I do option 1, I can divide the data into as many group as possible, right? not only 5? so then later I can have trainings for fold 0 to ... fold n? 

Bests,
Mehri

**Comment by FabianIsensee** [comment]
You can have as many splits as you want. I just updated the documentation in nnUNetTrainerV2 to make that clearer:

        The default split is a 5 fold CV on all available training cases. nnU-Net will create a split (it is seeded,
        so always the same) and save it as splits_final.pkl file in the preprocessed data directory.
        Sometimes you may want to create your own split for various reasons. For this you will need to create your own
        splits_final.pkl file. If this file is present, nnU-Net is going to use it and whatever splits are defined in
        it. You can create as many splits in this file as you want. Note that if you define only 4 splits (fold 0-3)
        and then set fold=4 when training (that would be the fifth split), nnU-Net will print a warning and proceed to
        use a random 80:20 data split.

**Comment by MehriB** [comment]
Thank you.

**Comment by adhusch** [comment]
Hi Fabian,
(hi Mehri),

briefly hooking in. Could please you elaborate 

> This is dirty because the experiment planning will have access to the metadata of all cases

In which way is the metadata utilized by experiment planning? Only for computing image size for resizing etc. or addtional other things?

Thank you
Best
Andy

**Comment by FabianIsensee** [comment]
Hi,
please have a look at the paper. It would take too long to explain it here.
Best,
Fabian
